<span style="font-size:10pt">&copy; 2019-2024 Michał Bukowski (m.bukowski@uj.edu.pl) ZBA WBBiB UJ</span>

---

## 1. Przygotowanie środowiska

<font size=2>Korzystając z magicznej komendy `!`, utwórz komendą Bash `mkdir` katalogi `dbs` i `output`. Po prostu podaj wszystkie ścieżki/nazwy katalogów oddzielone spacjami po wpisaniu samej komendy. Możesz skorzystać z paramteru `-p`, aby uniknąć pojawiania się błędu, gdy katalogi już istnieją (np. w trakcie ponownego uruchomienia kodu całego notatnika).
</font>

In [1]:
!mkdir -p db output

A subdirectory or file -p already exists.
Error occurred while processing: -p.
A subdirectory or file db already exists.
Error occurred while processing: db.
A subdirectory or file output already exists.
Error occurred while processing: output.


---

## 2. BLAST: tworzenie bazy danych


#### Narzędzie _makeblastdb_ i tworzenie bazy danych z wykorzystaniem przekierowania do _stdin_

<font size=2>Wiedząc, że w powłoce tekstowej Bash:

* wyjście standardowe (<i>standard output</i>, <i>stdout</i>) domyślnie jest tym co widzisz na ekranie
* _stdout_ z jednej aplikacji można przekierować do wejścia standardowego (<i>standard input</i>, <i>stdin</i>) do innej znakiem `|`,\
np.: `app1 | app2`
* aplikacja `cat plik1.txt plik2.txt ...` kopiuje do <i>stdout</i> zawartość pliku(ów)
* nazwy wielu plików o takim samym wzorcu nazwy można zastępować wyrażeniem zastępczym (<i>wildcard</i>),\
np.: `cat *.txt`
* aby utworzyć nukleotydową bazę danych narzędziem `makeblastdb` należy podać mu przynajmniej trzy parametry: `-in` ścieżkę do pliku z sekwencjami w formacie fasta, `-dbtype` <i>nucl</i> w przypadku sekwencji nukleotydowych i `-out` ścieżkę do katalogu wraz z prefiksem nazw plików tworzonej bazy danych
* parametr `-in` przyjmuje tylko jeden plik, można go pominąć a wtedy `makeblastdb` będzie odczytywało dane ze strumienia <i>stdin</i>

i korzystając z odpowiedniej `%%magicznej komendy` wpisz i uruchom w poniższej komórce kod Bash, który aplikacją `cat` wrzuci zawartość plików `*.fna` z katalogu `genomes` do <i>stdout</i> i przekieruje je do <i>stdin</i> `makeblastdb` celem utworzenia nukleotydowej bazy danych `genomes` w katalogu `dbs`
\
\
Następnie dopisz linię `ls -l dbs`, która wyświetli zawartość katalogu `dbs`

</font>

In [7]:
%%bash
cat genomes/*.fna | makeblastdb -dbtype nucl -out dbs/genomes -title "Nucleotide db"
ls -l dbs

#cat genomes/*.fna - wczytanie zawartości wszystkich plików .fna z katalogu genomes i wysłanie ich do stdout.
#| makeblastdb -dbtype nucl -out dbs/genomes - Przekierowanie danyhc z stdout do stdin narzędzia makeblastdb, tworząc bazę danych nukleotydową o nazwie genomes w katalogu dbs.
#ls -l dbs: Wyświetlanie szczegółowej listę zawartości katalogu dbs, aby sprawdzić, czy baza danych została utworzona poprawnie.




Building a new DB, current time: 10/28/2024 13:51:39
New DB name:   /mnt/c/Users/Acer/Desktop/bioinfa/bioinf ćw/cw 2/dbs/genomes
New DB title:  Nucleotide db
Sequence type: Nucleotide
Keep MBits: T
Maximum file size: 1000000000B
Adding sequences from FASTA; added 222 sequences in 0.338789 seconds.


total 6792
-rwxrwxrwx 1 root root   20480 Oct 28 13:51 genomes.ndb
-rwxrwxrwx 1 root root   43232 Oct 28 13:51 genomes.nhr
-rwxrwxrwx 1 root root    2764 Oct 28 13:51 genomes.nin
-rwxrwxrwx 1 root root    2672 Oct 28 13:51 genomes.not
-rwxrwxrwx 1 root root 6857061 Oct 28 13:51 genomes.nsq
-rwxrwxrwx 1 root root   16384 Oct 28 13:51 genomes.ntf
-rwxrwxrwx 1 root root     892 Oct 28 13:51 genomes.nto


---

## 3. BLAST: przeszukiwanie bazy danych
<font size=2>Następnie w kolejnej komórce wpisz i uruchom narzędzie `blastn`, podając jako sekwencję _query_ ścieżkę do pliku `input/gap.fna`, jako bazę danych, ścieżkę do tej dopiero utworzonej, a wyniki zapisz do pliku `output/blastn.txt`. Nie zapomnij o odpowiedniej `%%magicznej komendzie`, która pozwoli Ci uruchomić zawartość całej komórki jako kod Bash.
\
\
Następnie dopisz linię `head -25 output/blast.txt`, która wyświetli pierwszych 25 linii pliku z wynikami.
\
\
Na koniec przejdź do ekranu głównego Jupytera, wejdź do katalogu `output`, otwórz i przeglądnij całą zawartość pliku `blastn.txt`.</font>

In [9]:
%%bash
blastn -query input/gap.fna -db dbs/genomes -out output/blastn.txt
head -25 output/blastn.txt



BLASTN 2.12.0+


Reference: Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb
Miller (2000), "A greedy algorithm for aligning DNA sequences", J
Comput Biol 2000; 7(1-2):203-14.



Database: Nucleotide db
           222 sequences; 27,426,152 total letters



Query= gap

Length=1011
                                                                      Score     E
Sequences producing significant alignments:                          (Bits)  Value

UDMK01000001.1 Staphylococcus aureus strain EOE252 genome assembl...  1851    0.0  
UDIR01000001.1 Staphylococcus aureus strain EOE046 genome assembl...  1851    0.0  
UHCG01000001.1 Staphylococcus aureus strain NCTC9553 genome assem...  1851    0.0  
OGBR01000034.1 Staphylococcus aureus strain RH238 genome assembly...  1851    0.0  
FQKY01000001.1 Staphylococcus argenteus strain 3688STDY6125110 ge...  1757    0.0  


---

## 4. BLAST: otrzymywanie wyników w formacie tabelarycznym
<font size=2>Skpiuj kod z poprzedniej komórki i zmodyfikuj go tak, aby narzędzie `blastn` zapisało wyniki w formie tabelarycznej do pliku `output/blast.tsv` z tabulatorem jako sepratorem kolumn i z kolumnami `sseqid sstart send qcovs pident evalue`. Do jakich wartości odnoszą się te kolumny? Jeżeli nie wiesz, w terminalu tekstowym wpisz `blastn -help` i sprawdź.
\
\
Komendą `head` wyświetl tylko 5 pierwszych linii z pliku z wynikami.
\
\
Na koniec przejdź do ekranu głównego Jupytera, wejdź do katalogu `output`, otwórz i przeglądnij całą zawartość pliku `blastn.tsv`.

In [12]:
%%bash
blastn -query input/gap.fna -db dbs/genomes -out output/blastn.tsv -outfmt "6 sseqid sstart send qcovs pident evalue"
head -5 output/blastn.tsv



UDMK01000001.1	276652	277662	100	99.703	0.0
UDIR01000001.1	129950	128940	100	99.703	0.0
UHCG01000001.1	795488	796498	100	99.703	0.0
OGBR01000034.1	321008	322018	100	99.703	0.0
FQKY01000001.1	1106566	1105556	100	98.022	0.0


sseqid: ID dopasowanej sekwencji. \
sstart: Początek dopasowania na sekwencji odniesienia. \
send: Koniec dopasowania na sekwencji odniesienia. \
qcovs: Pokrycie sekwencji zapytania (%). \
pident: Procent identycznych dopasowań. \
evalue: Prawdopodobieństwo przypadkowego dopasowania (im mniejsze, tym lepiej).

---

## 5. Pandas: przetwarzanie wyników BLAST

<font size=2>Aby wykonać kolejne kroki, wykorzystaj prezentację z ostatniego wykładu jako pomoc.</font>

In [13]:
# zaimportuj moduł pandas pod nazwą 'pd'

import pandas as pd


In [27]:
# wczytaj wyniki z pliku output/balstn.tsv do obiektu typu DataFrame
# a następnie wyświetl tabelę

df = pd.read_csv(
    'output/blastn.tsv',
    names = 'sseqid sstart send qcovs pident evalue'.split(),
    sep = '\t' 
 )
df

sseqid   sstart     send  qcovs  pident  evalue
0  UDMK01000001.1   276652   277662    100  99.703     0.0
1  UDIR01000001.1   129950   128940    100  99.703     0.0
2  UHCG01000001.1   795488   796498    100  99.703     0.0
3  OGBR01000034.1   321008   322018    100  99.703     0.0
4  FQKY01000001.1  1106566  1105556    100  98.022     0.0
5  FQRR01000001.1  1121207  1120197    100  97.923     0.0
6      AE015929.1   553355   554365    100  89.229     0.0
7  FKIN01000005.1  1151570  1150579     98  86.847     0.0
8      LR134263.1  2257955  2256964     98  86.231     0.0
9  FMPF01000004.1   225112   226103     98  86.014     0.0

In [28]:
# utwórz nową tabelę poprzez zachowanie z wyjściowej tabeli
# wierszy gdzie wartość kolumny qcovs wynosi 100,
# wyświetl ją
df_qcovs_100=df[df['qcovs']==100]
df_qcovs_100




sseqid   sstart     send  qcovs  pident  evalue
0  UDMK01000001.1   276652   277662    100  99.703     0.0
1  UDIR01000001.1   129950   128940    100  99.703     0.0
2  UHCG01000001.1   795488   796498    100  99.703     0.0
3  OGBR01000034.1   321008   322018    100  99.703     0.0
4  FQKY01000001.1  1106566  1105556    100  98.022     0.0
5  FQRR01000001.1  1121207  1120197    100  97.923     0.0
6      AE015929.1   553355   554365    100  89.229     0.0

In [30]:
# zapisz tabelę z przefiltrowanymi wynikami
# do pliku output/filtered.tsv
# a następnie z ekranu głównego Jupyter Notebook
# wejdź do katalogu output i przeglądaj tak zapisane
# wyniki oraz porównaj je z zawartością pliku z
# wynikami wyjściowymi
df_qcovs_100.to_csv(
    'output/filtered.tsv',
    index = False,
    sep = '\t'
 )



---

## 6. Pandas: _DataFrame_ w trybie tekstowym

<font size=2>Uruchom terminal tekstowy, przejdź do aktualnego katalogu roboczego komendą Bash `cd ścieżka_katalogu`, uruchom tekstowy interpreter `ipython`, wczytaj tabelę z wyjściowymi wynikami z pliku `blastn.tsv` do struktury DataFrame i przeglądaj ją w trybie tekstowym.
\
\
Jeżeli nie pamiętasz ścieżki aktualnego katalogu roboczego, skorzystaj z magicznej komendy `%pwd`.</font>

   UDMK01000001.1   276652   277662  100  99.703  0.0\
0  UDIR01000001.1   129950   128940  100  99.703  0.0\
1  UHCG01000001.1   795488   796498  100  99.703  0.0\
2  OGBR01000034.1   321008   322018  100  99.703  0.0\
3  FQKY01000001.1  1106566  1105556  100  98.022  0.0\
4  FQRR01000001.1  1121207  1120197  100  97.923  0.0\
5      AE015929.1   553355   554365  100  89.229  0.0\
6  FKIN01000005.1  1151570  1150579   98  86.847  0.0\
7      LR134263.1  2257955  2256964   98  86.231  0.0\
8  FMPF01000004.1   225112   226103   98  86.014  0.0\